File: src/main/java/org/hisp/dhis/lib/expression/syntax/ParseContext.java
package org.hisp.dhis.lib.expression.syntax;

import org.hisp.dhis.lib.expression.ast.Node;
import org.hisp.dhis.lib.expression.ast.NodeType;
import org.hisp.dhis.lib.expression.ast.Position;

import java.util.function.Function;

/**
 * When parsing the expression grammar the context has two roles:
 *
 * 1. allow to emit nodes in the AST by using begin/end or add node methods
 * 2. lookup named building blocks by name to continue parsing
 *
 * @author Jan Bernitt
 */
public interface ParseContext extends FragmentContext
{
    /*
        Building the AST
     */

    void beginNode(NodeType type, Position start, String value, Node.Factory create );
    void endNode(NodeType type, Position end);

    /*
        Building the AST convenience methods
     */

    default void beginNode(NodeType type, Position start, String value ) {
        beginNode(type, start, value, null);
    }

    default void addNode(NodeType type, Position start, String value )
    {
        addNode(type, start, value, null);
    }

    default void addNode(NodeType type, Position start, String value, Node.Factory create )
    {
        beginNode( type,start, value, create );
        endNode(type, start == null ? null : start.offsetBy(value.length()));
    }
    default void addNode(NodeType type, Expr expr, Function<Expr, String> parse)
    {
        addNode(type, null, expr, parse);
    }
    default void addNode(NodeType type, Node.Factory factory, Expr expr, Function<Expr, String> parse)
    {
        Position start = expr.marker();
        int s = expr.position();
        try {
            addNode(type, start, parse.apply(expr), factory);
        } catch (RuntimeException ex)
        {
            expr.error(s, ex.getMessage());
        }
    }

}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/syntax/ParseContext.kt
package org.hisp.dhis.lib.expression.syntax

import org.hisp.dhis.lib.expression.ast.Node
import org.hisp.dhis.lib.expression.ast.NodeType
import org.hisp.dhis.lib.expression.ast.Position

/**
 * When parsing the expression grammar the context has two roles:
 *
 *
 * 1. allow to emit nodes in the AST by using begin/end or add node methods 2. lookup named building blocks by name to
 * continue parsing
 *
 * @author Jan Bernitt
 */
interface ParseContext : FragmentContext {
    /*
        Building the AST
     */
    fun beginNode(type: NodeType, start: Position?, value: String, create: Node.Factory?)
    fun endNode(type: NodeType, end: Position?)

    /*
        Building the AST convenience methods
     */
    fun beginNode(type: NodeType, start: Position?, value: String) {
        beginNode(type, start, value, null)
    }

    fun addNode(type: NodeType, start: Position?, value: String, create: Node.Factory? = null) {
        beginNode(type, start, value, create)
        endNode(type, start?.offsetBy(value.length))
    }

    fun addNode(type: NodeType, expr: Expr, parse: (Expr) -> String) {
        addNode(type, null, expr, parse)
    }

    fun addNode(type: NodeType, factory: Node.Factory?, expr: Expr, parse: (Expr) -> String ) {
        val start = expr.marker()
        val s = expr.position()
        try {
            addNode(type, start, parse(expr), factory)
        } catch (ex: RuntimeException) {
            expr.error(s, ex.message)
        }
    }
}
